File: 2022/w14/mars_rover_kata_go/direction.go
package kata

type direction rune

func newDirection() direction {
	return N
}

func (d direction) left() direction {
	return lefts[d]
}

func (d direction) right() direction {
	return rights[d]
}

func (d direction) String() string {
	return string(d)
}

var (
	lefts = map[direction]direction{
		N: W,
		E: N,
		S: E,
		W: S,
	}

	rights = map[direction]direction{
		N: E,
		E: S,
		S: W,
		W: N,
	}
)

const (
	N direction = 'N'
	E direction = 'E'
	S direction = 'S'
	W direction = 'W'
)

File: 2022/w21/mars_rover_kata_js/src/direction.js
'use strict'

const N = 'N'
const E = 'E'
const S = 'S'
const W = 'W'

const lefts = new Map([
  [N, W],
  [E, N],
  [S, E],
  [W, S]
])

const rights = new Map([
  [N, E],
  [E, S],
  [S, W],
  [W, N]
])

class Direction {
  static create(initialDirection = N) {
    return new Direction(initialDirection)
  }

  constructor(initialDirection) {
    Object.defineProperty(this, 'value', {
      value: initialDirection,
      writable: false,
      enumerable: true,
      configurable: true
    })
  }

  direction() {
    return this.value
  }

  left() {
    return lefts.get(this.value)
  }

  right() {
    return rights.get(this.value)
  }

  toString() {
    return this.value
  }
}

module.exports = { Direction, N, E, S, W }
